### ISSUE TITLE
Concurrent state channels for a user don't work correctly

### ISSUE THREAD

**Issue by ycryptx** [ycryptx/recursive-zk-lightning-network#5]
Currently, when a user opens concurrent channels for a given token, each channel can use up the full amount deposited by the user. This will cause some channels to fail indeterminately -- something that the protocol should never allow.

**Comment by HyunggyuJang** [comment]
## Native solution
Just use Lightning contract per channel ;)
## Serious solution?
### Time lock should be linked per channel
currently it is global: users deposit tokens when entering contract, not channel
### Make the channel initialization explicitly
I.e., make the channel opening to be an on-chain activity: currently users can open new channel off-chain; so there is no way to check whether the channel opening is valid, i.e., there is no double spending.

Specifically, we can do following to solve this issue:

1. On channel opening, subtract the balance user want to lock to given channel,
2. On channel closing, add remaining balance to the user's deposit

This way, each channel can be started and ended without interfering other channels.

**Comment by HyunggyuJang** [comment]
> I think your `serious solution` is right. Deposits should be user-to-user not user-to-contract and in doing so there will be a dedicated channel with its own balance between 2 users. If the deposit already instantiate the two way channel between two users there's only need for a single on-chain transaction to open a channel and start using it

Yes :) I fixed the first solution, which meant to be "naive", rather than native, sorry for the misleading!
